iFixit/react-commerce
# Upgrade to next 13

closes #1595 

This pull request includes the following updates:

### Next Image

The `next/future/image` is now the default image component as per [Next.js 13 documentation](https://nextjs.org/docs/pages/building-your-application/upgrading/version-13#image-component). Some props were removed which required changes to both usages of the `next/image` component and derived components like `ResponsiveImage`. A few style fixes were needed due to markup differences.

### Next Links

The new link component doesn't require a nested anchor tag anymore. Whenever possible, the new behavior was adopted. The `legacyBehaviour` prop is available and has been used to reduce the number of changes in this PR for components that could've needed more intricate style/markup refactoring.

### Next Script

Scripts marked as `beforeInteractive` were moved to `_document.tsx` as suggested in [Next.js documentation](https://nextjs.org/docs/messages/no-before-interactive-script-outside-document).

### Next Transpile Modules

Next.js now has its own transpile option, so the `next-transpile-modules` package is no longer needed. For more information, check [here](https://github.com/martpie/next-transpile-modules/issues/280#issuecomment-1292615030).

### Sentry Edge Config

A Sentry edge config was added to address warnings about missing Sentry config files and to prepare for experiments with edge middlewares.

### SSR Failure for Next Router

A workaround was added to prevent SSR failure in Next.js 13. Refer to these discussions for more information: [issue #5650](https://github.com/algolia/instantsearch/issues/5650) and [issue #5475](https://github.com/algolia/instantsearch/issues/5475#issuecomment-1425897322).

### MSW Issue

The MSW mocking utility is not fully compatible with Next 13, as discussed [here](https://github.com/mswjs/msw/issues/1610#issuecomment-1537208102). Since MSW will likely be replaced by other mocking strategies, problematic tests have been temporarily skipped.
The reasons we suggest prioritizing an upgrade over exploring new testing mechanisms are that Next.js 13 introduces a significant change with server components. 
Delaying the adoption of this paradigm might be disadvantageous. During these weeks, several core aspects are being discussed, such as translations management, performance improvement, and A/B testing—all of which could benefit from the enhanced functionality offered by the new framework.
For additional details see [here](https://github.com/iFixit/react-commerce/pull/1726#discussion_r1232395722)

### Minor Updates

- A missing hook dependency was added to `InstantSearchProvider.tsx`.

### QA

1. Visit Vercel preview
2. Verify functionality across the various page types 
3. Take into account that several playwright test were skipped since MSW is not yet compatible with Next.js 13 - so there might be a few more scenarios to verify manually.
